Fixes for Python3, switched to pcryptodome

BenWoodford opened the pull request:
This includes numerous fixes for Python3 from @tomhillable and @filcole, which I've merged (sometimes manually...) into my own fork - but I feel like I should PR them. I hope the others don't mind, but right now this lib doesn't actually work with later versions of Python.

tomhillable committed:
Fix import of responses

tomhillable committed:
Follow PEP8 indentation guidelines

tomhillable committed:
Fix relative import it __init__.py

tomhillable committed:
Fix "TypeError: Object of type bytes is not JSON serializable"

tomhillable committed:
PEP8 styling changes

tomhillable committed:
Fix KeyError with PriceSimulatorTotalInfo responses

tomhillable committed:
Fix "TypeError: float() argument must be a string or a number, not NoneType"

tomhillable committed:
Fix "KeyError: ElectricBill"

Commit:
Attempting to fix pycrypto issues

Commit:
Added some Py3 fixes from...

https://github.com/filcole/pycarwings2

Commit:
Switch to pycryptodome

tomhillable commented:
Thanks @BenWoodford, fine by me. 👍

filcole commented:
I like it.  Thanks @BenWoodford

filcole committed:
Handle missing "BatteryStatusRecords"

In case of a long time since requesting an update from the car the Nissan servers return just `{ "status": 200 }`.  Fix to avoid crash in this
instance.

Commit:
Removing range from start response

benroy73 committed:
fix spelling

benroy73 committed:
following pep8 style guide

benroy73 committed:
following pep8 style guide

benroy73 committed:
add pytest and flake8 all the tests can be run with "./setup.py test"

benroy73 committed:
updated to work with python 3.6 and 2.7

benroy73 committed:
update travis ci and fix test in examples

benroy73 committed:
python 3.4 needs this decoded to a str

benroy73 committed:
test with python 2.7, 3.4 and 3.6

benroy73 committed:
removed 3.4 since tests are not working

BenWoodford commented:
Pinging @jdhorne as I don't think PRs fire a Github notification

jdhorne commented:
Hi @BenWoodford; does this break compatibility with Python 2? I know some folks are still using that. Just want to make sure I know what I'm doing by approving.

Commit:
Merge branch 'master' into master

BenWoodford commented:
Honestly? No idea. I wasn't aware people were even actively using Python 2 for new projects.

I would _assume_ so. This is mostly to fix stuff like tabs being used instead of spaces, etc. However I don't know if `pycryptodome` supports Python 2.

Commit:
Added Python 3.4 to travis

BenWoodford commented:
Travis builds so I'd assume all is well. I've just added Python 3.4 as a second test too, which passes. The only major change that could have affected it is the use of a different crypto library.

filcole committed:
Don't crash when battery status not returned 

Fix in this repo per issue https://github.com/jdhorne/pycarwings2/issues/29

glynhudson committed:
typo

filcole committed:
Add example ELECTRIC_WAVE_ABNORMAL response

filcole committed:
Don't crash if cruising ranges aren't returned

RemoteACRecordsRequest.php does not always return cruising ranges.

filcole committed:
Merge pull request #1 from filcole/MissingCruisingRange

Missing cruising range and ELECTRIC_WAVE_NORMAL

filcole committed:
Merge in latest from BenWoodford

filcole committed:
Merge in GlynHudson

filcole committed:
Merge branch 'glynhudson-patch-1' into HomeAssistant

filcole committed:
Fix spelling of exemples directory

filcole committed:
Update README.md to HomeAssistant branch

filcole committed:
Handle missing "BatteryStatusRecords"

In case of a long time since requesting an update from the car the
Nissan servers return just `{ "status": 200 }`. Fix to avoid crash in
this instance.

filcole committed:
Merge branch 'master' of https://github.com/benroy73/pycarwings2 into benroy73-master

filcole committed:
Merge branch 'benroy73-master' into HomeAssistant

filcole committed:
Use 4 char spaces instead of tabs

filcole committed:
Re-apply handle missing "BatteryStatusRecords"

In case of a long time since requesting an update from the car the
Nissan Servers return just { "start": 200 }. Fix to avoid in
this instance.

filcole committed:
Re add example ELECTRIC_WAVE_ABNORMAL response

filcole committed:
Re add Don't crash if cruising ranges aren't returned

filcole committed:
Merge pull request #5 from filcole/HomeAssistant

Home assistant

Commit:
Merge pull request #2 from filcole/patch-2

Don't crash when battery status not returned

Commit:
Merge pull request #1 from filcole/patch-1

Handle missing "BatteryStatusRecords"

Commit:
Fixing base URL

As per https://github.com/jdhorne/pycarwings2/issues/35

Commit:
Fixed lack of trailing slash in base URL

Commit:
Trying a different URL

filcole committed:
Fix missing colon

filcole committed:
Fix imports so that can run test

filcole committed:
Ignore virtualenv directorires

filcole committed:
Update setup.py for packaging

filcole committed:
Fix comment referring to python2

filcole committed:
Fixes for flake. Ignore W504 warning

filcole committed:
Fixes for flake8

filcole committed:
Update BASE_URL

filcole committed:
Ignore .pytest_cache

filcole committed:
Add MANIFEST.in

filcole committed:
Update encryption strings and improve example script

filcole committed:
Update example to include region and fix URL in README.md

filcole committed:
Update control_climate.py example to use config file. Remove log file

filcole committed:
Read region from config for get-leaf-location example.

filcole committed:
Add CarwingsError to exports and fix test_bad_password()

filcole committed:
Fixes for Flake8

filcole committed:
Fixes for Flake8

filcole committed:
Add initial tox.ini config file, just for py35

filcole committed:
Fix typo

filcole committed:
Update README to comment on nissan server behaviour

filcole committed:
Try adding setuptools to travis

filcole committed:
Add python to top of script for Travis

filcole committed:
Update README.md

filcole committed:
Add Python3.7 for travis and don't install setuptools

filcole committed:
Remove 3.7 from travis.  Will revisit

filcole committed:
Fix PyPi classifiers

filcole committed:
Raise exception instead of crashing if HTTP response not received.

filcole committed:
Add get latest battery example

filcole committed:
Handle divide by zero when battery_capacity=0

filcole commented:
I've used a new URL, the same one used by dartcarwings, and included in a PyPi release.  Please see my fork  at https://github.com/filcole/pycarwings2 

I've also released to PyPi so that it can be more easily used by other projects (e.g. HomeAssistant)

filcole committed:
Fix license reported by pip

filcole committed:
Add start-charging.py example, comment that only command is sent.

filcole committed:
Fix incorrect quotes and missing commas in Climate Control error example

filcole committed:
Handle LatestClimateControl not returning any data

filcole committed:
Fixes picked up by tox

filcole committed:
Release 2.3

filcole committed:
Update install instruction since now in PyPi

remyderuysscher commented:
> 
> 
> I've used a new URL, the same one used by dartcarwings, and included in a PyPi release. Please see my fork at https://github.com/filcole/pycarwings2
> 
> I've also released to PyPi so that it can be more easily used by other projects (e.g. HomeAssistant)

@filcole  would you be so kind to fix the Home Assistant implementation for the Nissan Leaf? @BenWoodford doesn't have any time. Your implementation works wonderful ! 

Thanks.

filcole commented:
@remyderuysscher see pull request https://github.com/home-assistant/home-assistant/pull/19786 for the ongoing effort to create an integration with homeassistant.  Any help appreciated. :)

remyderuysscher commented:
> 
> 
> @remyderuysscher see pull request [home-assistant/home-assistant#19786](https://github.com/home-assistant/home-assistant/pull/19786) for the ongoing effort to create an integration with homeassistant. Any help appreciated. :)

@filcole Sure, let me know how I can help.

filcole committed:
Fix incorrect assert in connection tests

filcole committed:
Fix handling of missing climate info, add tests and example

filcole committed:
Version 2.4

filcole committed:
Increase timeout to 30 seconds in example

filcole committed:
Set battery_percent if have SOC. Note that BatteryStatusCheck now always returns 0. Handle invalid JSON.

filcole committed:
Flake8 fixes

filcole committed:
Version 2.5

filcole committed:
Ignore Scripts folder created for Windows virtenvs

filcole committed:
Remove requirement for specific version of pycrytodome

filcole committed:
Version 2.6

filcole committed:
battery_percent should be a float

filcole committed:
Version 2.7

filcole committed:
Older (24kWh?) leafs return out of 12 bars, even if degrated to 11 or lower.

filcole committed:
Add tests for battery_percent

filcole committed:
Add test for 24kWh 100% 1 degredation

filcole committed:
Version 2.8

filcole committed:
Add tests for climate responses without CruiseRangeAcOn/Off

jrester committed:
API changes 25 July 2019

jrester committed:
Add method update_battery_status and update get-leaf-info.py accordingly

filcole committed:
Move update_battery_status to examples

filcole committed:
Add slightly changed BatteryStatusCheckResultRequest.php response

filcole committed:
Version 2.9

filcole committed:
Fix whitespace issues

filcole committed:
Merge pull request #16 from filcole/dev

Fix 25 July API changes.

Commit:
Updated travellist to response

filcole committed:
Merge pull request #18 from jannetak/master

Updated travellist to response

Kalkran committed:
Suppress the user-agent from newer versions of urllib3

Kalkran committed:
fixed indentation

Kalkran committed:
PEP8 Indentation again, I don't do this often

filcole committed:
Set identifiable User-Agent

filcole committed:
Use python 3.7 for tox

filcole committed:
Remove obsolete setup.py option

filcole committed:
Version 2.10

filcole committed:
Merge pull request #25 from filcole/dev

Version 2.10

BenWoodford committed:
Merge branch 'master' of https://github.com/filcole/pycarwings2
